data-and-programming.Rmd

```diff
@@ -182,93 +204,175 @@ We see that when a function like `log()` is called on a vector `x`, a vector is
+Here we see that using the `sum()` function on a vector of logical `TRUE` and `FALSE` values that is the result of `x > 3` results in a numeric result. `R` is first automatically coercing the logical to numeric where `TRUE` is `1` and `FALSE` is `0`. This coercion from logical to numeric happens for most mathematical operations. Under the hood, something akin to the following is happening.
+
+```{r}
+x = c(1, 3, 5, 7, 8, 9)
+temp = as.numeric(x > 3)
+sum(temp)
+```
+
+The `which` function is useful for extracting out all of the index values from a vector where a logical evaluates to `TRUE`. 
 
 ```{r}
+x = c(1, 3, 5, 7, 8, 9)
 which(x > 3)
+```
+
+These indicies can, in turn, be used to lookup values in the original vector.
+
+```{r}
+x = c(1, 3, 5, 7, 8, 9)
 x[which(x > 3)]
+```
+
+Often times we want to know the largest element in a vector. The `max` function is useful for that.
 
+```{r}
+x = c(1, 3, 5, 7, 8, 9)
 max(x)
-which(x == max(x))
-which.max(x)
 ```
 
-### More Vectorization
+This can be combined with `which` to find index where the maximum value occurs.
 
 ```{r}
 x = c(1, 3, 5, 7, 8, 9)
-y = 1:100
+which(x == max(x))
 ```
 
+Alternatively, the `which.max` returns the index with the maximum value straight away.
+
 ```{r}
+x = c(1, 3, 5, 7, 8, 9)
+which.max(x)
+```
+
+Similar functions `min` and `which.min` are also available.
+
+### More Vectorization
+
+In addition to coercing arguments to the right type, `R` also expands vectors to the correct length. We have seen this already. For example, the following operations are equivalent.
+
+```{r results = 'hold'}
+x = c(1, 3, 5, 7, 8, 9)
 x + 2
+x + c(2, 2, 2, 2, 2, 2)
 x + rep(2, 6)
 ```
 
-```{r}
+These operations are also equivalent.
+
+```{r results = 'hold'}
+x = c(1, 3, 5, 7, 8, 9)
 x > 3
+x > c(3, 3, 3, 3, 3, 3)
 x > rep(3, 6)
 ```
 
+What about when the vector lengths are different? `R` handles this too but issues a warning when the lengths are not multiples of one another.
+
 ```{r}
+x = c(1, 3, 5, 7, 8, 9)
+y = 1:100
 x + y
+```
+
+Again, the warning is because of the lengths of the vectors being used are not multiples of one another.
+
+```{r results = 'hold'}
 length(x)
 length(y)
 length(y) / length(x)
+```
+
+Likewise, this will generate a warning.
+
+```{r}
 (x + y) - y
 ```
 
+If the lengths are multiples of each other, no warning is issued. For example:
+
 ```{r}
+x = c(1, 3, 5, 7, 8, 9)
 y = 1:60
 x + y
+```
+
+No warning because `y` is an even multiple of `x`.
+
+```{r}
 length(y) / length(x)
 ```
 
+The expansion can happen on both the left-hand side and right-hand side of the operation.
+
 ```{r}
+x = c(1, 3, 5, 7, 8, 9)
+y = 1:60
 rep(x, 10) + y
 ```
 
-```{r}
+Using the `all` or `identical` functions we can check that, indeed, both left-hand side expansion and right-hand side expansion are the same.
```

Comment:
This is a boo-boo with the text. In particular, @daviddalpiaz `identical()` => `all.equal()`. There are _very_ few cases where `identical()` is appropriate. c.f. [What every computer scientist should know about floating-point arithmetic
](http://www.itu.dk/~sestoft/bachelor/IEEE754_article.pdf) (Goldberg, 1991).

Reply:
I'd argue that sometimes, it is useful to at least know what `identical()` does. It even sort of comes up in the course. I should probably explain the difference between `all.equal()` and `identical()` though. Again, how's your `R` textbook coming along?

Reply:
Which is more appropriate I think depends on the domain.  For numerical stuff, sure, all.equal() would be the way to go.  Systems level serialization, you almost certainly want bit-wise identical behavior.  But perhaps more emphasis could be drawn, including a reference to your suggested link.